### ISSUE TITLE
The `regex-shorthand` rule has a false-negative

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#157]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2430%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/157)

This doesn't cause a lint error:

```js
const foo = /^by @([a-zA-Z0-9-]+)/
```

Even though it could be:

```js
const foo = /^by @([a-zA-Z\d-]+)/
```

I think we might need to use some more advanced detection. Maybe [`regexp-tree`](https://github.com/DmitrySoshnikov/regexp-tree). Although, it has an [optimizer](https://github.com/DmitrySoshnikov/regexp-tree#using-optimizer-api) API already, which I think already handles this and many other cases.

// @SamVerschueren

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $30.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($30.00)

### Submitted pull Requests
- [#261 regex-shorthand: new option to fix #157](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/261)
- [#437 update  to use  for regex literals](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/437)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by IssuehuntBot** [comment]
@issuehunt has funded $30.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/157) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/157).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/users/44827199/repositories) on IssueHunt to raise funds.

**Comment by sindresorhus** [comment]
This is up for grabs again. If anyone wants to tackle this, you can probably continue where https://github.com/sindresorhus/eslint-plugin-unicorn/pull/261 left off.
